// @progress/kendo-angular-grid/dist/es2015/excel/excel-command.directive.js
import { Directive, HostListener, HostBinding } from '@angular/core';
import { ExcelService } from './excel.service';
/**
 * Represents the **Export to Excel** command of the Grid.
 *
 * You can apply this directive to any `button` element inside a
 * [`ToolbarTemplate`]({% slug api_grid_commandcolumncomponent %}).
 *
 * When the user clicks a button associated with the directive, the
 * [`excelExport`]({% slug api_grid_gridcomponent %}#toc-excelexport) event
 * fires. For more information, refer to article on [exporting the Grid to Excel]({% slug excelexport_grid %}).
 *
 * @example
 * ```ts-no-run
 * <kendo-grid>
 *      <ng-template kendoGridToolbarTemplate>
 *          <button kendoGridExcelCommand>Export to PDF</button>
 *      </ng-template>
 *      <kendo-grid-excel fileName="Grid.xlsx">
 *      </kendo-grid-excel>
 * </kendo-grid>
 * ```
 */
export class ExcelCommandDirective {
    constructor(excelService) {
        this.excelService = excelService;
    }
    /**
     * @hidden
     */
    click() {
        this.excelService.exportClick.emit();
    }
    /**
     * @hidden
     */
    get buttonClass() {
        return true;
    }
    /**
     * @hidden
     */
    get excelClass() {
        return true;
    }
}
ExcelCommandDirective.decorators = [
    { type: Directive, args: [{
                selector: '[kendoGridExcelCommand]'
            },] },
];
/** @nocollapse */
ExcelCommandDirective.ctorParameters = () => [
    { type: ExcelService, },
];
ExcelCommandDirective.propDecorators = {
    'click': [{ type: HostListener, args: ['click',] },],
    'buttonClass': [{ type: HostBinding, args: ['class.k-button',] },],
    'excelClass': [{ type: HostBinding, args: ['class.k-grid-excel',] },],
};
