// @progress/kendo-angular-inputs/dist/es2015/slider/slider-ticks.component.js
import { Component, Inject, Input, Optional, ViewChildren, HostBinding } from '@angular/core';
import { RTL } from '@progress/kendo-angular-l10n';
import { SliderUtil as Util } from '@telerik/kendo-inputs-common';
/**
 * @hidden
 */
class SliderTick {
    constructor(value) {
        this.value = value;
        this.classes = {
            'k-tick': true
        };
    }
}
/* tslint:disable:component-selector */
/**
 * @hidden
 */
export class SliderTicksComponent {
    constructor(rtl) {
        this.rtl = rtl;
        this.wrapperClasses = 'k-reset k-slider-items';
        this.ticks = [];
    }
    ngOnChanges(_) {
        this.createTicks();
    }
    createTicks() {
        const count = Util.calculateTicksCount(this.max, this.min, this.step);
        const tickValueProps = {
            min: this.min,
            max: this.max,
            smallStep: this.step
        };
        let result = [];
        for (let i = 0; i < count; i++) {
            result.push(new SliderTick(Util.calculateValueFromTick(i, tickValueProps)));
        }
        if (this.rtl || this.vertical) {
            result = result.reverse();
        }
        if (result.length > 0) {
            Object.assign(result[0].classes, this.endTickClasses(true));
            Object.assign(result[result.length - 1].classes, this.endTickClasses(false));
        }
        this.ticks = result;
    }
    endTickClasses(first) {
        return {
            'k-first': (first && !this.vertical) || (!first && this.vertical),
            'k-last': (!first && !this.vertical) || (first && this.vertical)
        };
    }
}
SliderTicksComponent.decorators = [
    { type: Component, args: [{
                selector: '[kendoSliderTicks]',
                template: `
    <li #tickElement *ngFor="let tick of ticks;"
        [ngClass]="tick.classes"
        title="{{ tickTitle(tick.value) }}"
        role="presentation"
     >&nbsp;</li>
  `
            },] },
];
/** @nocollapse */
SliderTicksComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [RTL,] },] },
];
SliderTicksComponent.propDecorators = {
    'wrapperClasses': [{ type: HostBinding, args: ['class',] },],
    'tickTitle': [{ type: Input },],
    'vertical': [{ type: Input },],
    'step': [{ type: Input },],
    'min': [{ type: Input },],
    'max': [{ type: Input },],
    'tickElements': [{ type: ViewChildren, args: ['tickElement',] },],
};
