// @progress/kendo-angular-treeview/dist/es/checkbox/checkbox.component.js
import { Component, ElementRef, EventEmitter, Input, Output, HostBinding, Renderer2 } from '@angular/core';
import { guid } from '../utils';
/**
 * @hidden
 *
 * Represents the CheckBox component of the Kendo UI TreeView for Angular.
 *
 */
var CheckBoxComponent = /** @class */ (function () {
    function CheckBoxComponent(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        /**
         * Specifies the [`id`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id) of the component.
         */
        this.id = "_" + guid();
        /**
         * Specifies the checked state of the component.
         *
         * The available options are:
         * - `none`&mdash;Is not checked.
         * - `indeterminate`&mdash;Sets the indeterminate state.
         * - `checked`&mdash;Sets the checked state.
         */
        this.checkState = 'none';
        /**
         * Specifies the [`tabindex`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex) of the component.
         */
        this.tabindex = 0;
        /**
         * Fires when the user changes the check state of the component.
         */
        this.checkStateChange = new EventEmitter();
    }
    Object.defineProperty(CheckBoxComponent.prototype, "classWrapper", {
        //XXX: implement ComponentValueAccessor
        //XXX: focus/blur methods
        get: function () { return true; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckBoxComponent.prototype, "indeterminate", {
        get: function () {
            return this.checkState === 'indeterminate';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckBoxComponent.prototype, "checked", {
        get: function () {
            return this.checkState === 'checked';
        },
        enumerable: true,
        configurable: true
    });
    CheckBoxComponent.prototype.ngOnInit = function () {
        this.renderer.removeAttribute(this.element.nativeElement, "tabindex");
    };
    CheckBoxComponent.prototype.handleChange = function (e) {
        this.checkStateChange.emit(e.target.checked ? 'checked' : 'none');
    };
    CheckBoxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'kendo-checkbox',
                    template: "\n        <input\n            class=\"k-checkbox\"\n            type=\"checkbox\"\n            [id]=\"id\"\n            [checked]=\"checked\"\n            [indeterminate]=\"indeterminate\"\n            [tabindex]=\"tabindex\"\n            (change)=\"handleChange($event)\"\n        />\n        <label\n            class=\"k-checkbox-label\"\n            tabindex=\"-1\"\n            [for]=\"id\"\n        >{{labelText}}</label>\n    "
                },] },
    ];
    /** @nocollapse */
    CheckBoxComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
    ]; };
    CheckBoxComponent.propDecorators = {
        'classWrapper': [{ type: HostBinding, args: ['class.k-checkbox-wrapper',] },],
        'id': [{ type: Input },],
        'checkState': [{ type: Input },],
        'labelText': [{ type: Input },],
        'tabindex': [{ type: Input },],
        'checkStateChange': [{ type: Output },],
    };
    return CheckBoxComponent;
}());
export { CheckBoxComponent };
