// @progress/kendo-angular-upload/dist/es2015/file-list-item-base.js
import util from "./util";
/**
 * @hidden
 */
export class FileListItemBase {
    constructor(uploadService) {
        this.uploadService = uploadService;
        this.progressComplete = 0;
    }
    subscribeUploadProgress(uploadProgressHandler) {
        this.uploadProgressSubscription = this.uploadService.uploadProgressEvent.subscribe(uploadProgressHandler);
    }
    fileHasValidationErrors(file) {
        return util.fileHasValidationErrors(file);
    }
    filesHaveValidationErrors(files) {
        return util.filesHaveValidationErrors(files);
    }
    ngOnDestroy() {
        this.uploadProgressSubscription.unsubscribe();
    }
    getFileValidationMessage(file) {
        let validationMessage;
        if (file.validationErrors && file.validationErrors.length > 0) {
            validationMessage = this.localization.get(file.validationErrors[0]);
        }
        return validationMessage;
    }
    getInvalidFilesMessage() {
        return this.localization.get('invalidFiles');
    }
    getTotalFilesSizeMessage(files) {
        return util.getTotalFilesSizeMessage(files);
    }
}
