// @progress/kendo-angular-upload/dist/es2015/upload-status-total.component.js
import { Component, Input } from '@angular/core';
import { FileState } from './file-info';
import { LocalizationService } from '@progress/kendo-angular-l10n';
/**
 * @hidden
 */
export class UploadStatusTotalComponent {
    constructor(localization) {
        this.localization = localization;
    }
    ngDoCheck() {
        let currentlyUploading = this.fileList.hasFileWithState([FileState.Uploading]);
        if (this.isUploading !== currentlyUploading) {
            this.isUploading = currentlyUploading;
            this.isFailed = this.fileList.hasFileWithState([FileState.Failed]);
            this.statusText = this.isUploading ? this.localization.get('headerStatusUploading')
                : this.localization.get('headerStatusUploaded');
        }
    }
    /**
     * @hidden
     */
    setStatusClasses() {
        return {
            "k-i-checkmark": !this.isUploading && !this.isFailed,
            "k-i-exception": !this.isUploading && this.isFailed,
            "k-i-loading": this.isUploading,
            "k-icon": true
        };
    }
}
UploadStatusTotalComponent.decorators = [
    { type: Component, args: [{
                selector: 'kendo-upload-status-total',
                template: `
        <strong
          class="k-upload-status k-upload-status-total">
          <span [ngClass]="setStatusClasses()"></span>
          {{statusText}}
        </strong>
    `
            },] },
];
/** @nocollapse */
UploadStatusTotalComponent.ctorParameters = () => [
    { type: LocalizationService, },
];
UploadStatusTotalComponent.propDecorators = {
    'fileList': [{ type: Input },],
};
