// @progress/kendo-angular-upload/dist/npm/file-select.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var upload_service_1 = require("./upload.service");
var navigation_service_1 = require("./navigation.service");
var util_1 = require("./util");
var validation_util_1 = require("./validation-util");
/**
 * @hidden
 */
var FileSelectDirective = /** @class */ (function () {
    function FileSelectDirective(uploadService, navigation, el) {
        this.uploadService = uploadService;
        this.navigation = navigation;
        this.type = "file";
        this.autocomplete = "off";
        this.tabIndex = -1;
        this.element = el;
    }
    Object.defineProperty(FileSelectDirective.prototype, "nameAttribute", {
        get: function () {
            return this.async.saveField;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FileSelectDirective.prototype, "multipleAttribute", {
        get: function () {
            return this.multiple ? "multiple" : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FileSelectDirective.prototype, "dirAttribute", {
        get: function () {
            return this.dir;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FileSelectDirective.prototype, "disabledAttribute", {
        get: function () {
            return this.disabled ? "true" : null;
        },
        enumerable: true,
        configurable: true
    });
    FileSelectDirective.prototype.onInputChange = function (event) {
        var _this = this;
        var ua = navigator.userAgent;
        var chrome = /(chrome)[ \/]([\w.]+)/i;
        var safari = /(webkit)[ \/]([\w.]+)/i;
        var selectedFiles = util_1.default.inputFiles(event.target);
        selectedFiles = util_1.default.assignGuidToFiles(selectedFiles, !this.async.batch);
        validation_util_1.default.validateFiles(selectedFiles, this.restrictions);
        if (!this.multiple) {
            this.uploadService.clearFiles();
        }
        this.uploadService.addFiles(selectedFiles, this.async);
        /*
        Chrome and Internet Explorer do not trigger a `change` event
        when a file with the same name is selected a number of consecutive times.
        As a workaround, clear the input value after handling the file.
        */
        var native = this.element.nativeElement;
        if (!(!ua.match(chrome) && ua.match(safari))) {
            native.type = "";
            native.type = "file";
        }
        setTimeout(function () {
            _this.navigation.focusedIndex = -1;
        });
    };
    FileSelectDirective.decorators = [
        { type: core_1.Directive, args: [{
                    selector: '[kendoFileSelect]'
                },] },
    ];
    /** @nocollapse */
    FileSelectDirective.ctorParameters = function () { return [
        { type: upload_service_1.UploadService, },
        { type: navigation_service_1.NavigationService, },
        { type: core_1.ElementRef, },
    ]; };
    FileSelectDirective.propDecorators = {
        'async': [{ type: core_1.Input },],
        'dir': [{ type: core_1.Input },],
        'disabled': [{ type: core_1.Input },],
        'multiple': [{ type: core_1.Input },],
        'restrictions': [{ type: core_1.Input },],
        'type': [{ type: core_1.HostBinding, args: ["attr.type",] },],
        'autocomplete': [{ type: core_1.HostBinding, args: ["attr.autocomplete",] },],
        'tabIndex': [{ type: core_1.HostBinding, args: ["attr.tabIndex",] },],
        'nameAttribute': [{ type: core_1.HostBinding, args: ["attr.name",] },],
        'multipleAttribute': [{ type: core_1.HostBinding, args: ["attr.multiple",] },],
        'dirAttribute': [{ type: core_1.HostBinding, args: ["attr.dir",] },],
        'disabledAttribute': [{ type: core_1.HostBinding, args: ["attr.disabled",] },],
        'onInputChange': [{ type: core_1.HostListener, args: ["change", ["$event"],] },],
    };
    return FileSelectDirective;
}());
exports.FileSelectDirective = FileSelectDirective;
