// @progress/kendo-angular-upload/dist/npm/upload-events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * @hidden
 */
var PreventableEvent = /** @class */ (function () {
    function PreventableEvent() {
        this.prevented = false;
    }
    /**
     * Prevents the default action for a specified event.
     * In this way, the source component suppresses the built-in behavior that follows the event.
     */
    PreventableEvent.prototype.preventDefault = function () {
        this.prevented = true;
    };
    /**
     * If the event is prevented by any of its subscribers, returns `true`.
     *
     * @returns `true` if the default action was prevented. Otherwise, returns `false`.
     */
    PreventableEvent.prototype.isDefaultPrevented = function () {
        return this.prevented;
    };
    return PreventableEvent;
}());
exports.PreventableEvent = PreventableEvent;
/**
 * Fires when the user cancels the process of uploading a file or a batch of files.
 *
 * {% embed_file events/cancel/upload.component.ts %}
 * {% embed_file shared/app.component.ts %}
 * {% embed_file shared/main.ts hidden %}
 *
 * Arguments for the `cancel` event:
 */
var CancelEvent = /** @class */ (function () {
    /**
     * Constructs the event arguments for the `cancel` event.
     * @param files - The list of the files that were going to be uploaded.
     */
    function CancelEvent(files) {
        this.files = files;
    }
    return CancelEvent;
}());
exports.CancelEvent = CancelEvent;
/**
 * Fires when the **Clear** button is clicked. At this point, the selected files are about to be cleared.
 *
 * {% embed_file events/clear/upload.component.ts %}
 * {% embed_file shared/app.component.ts %}
 * {% embed_file shared/main.ts hidden %}
 *
 */
var ClearEvent = /** @class */ (function (_super) {
    tslib_1.__extends(ClearEvent, _super);
    /**
     * Constructs the event arguments for the `clear` event.
     */
    function ClearEvent() {
        return _super.call(this) || this;
    }
    return ClearEvent;
}(PreventableEvent));
exports.ClearEvent = ClearEvent;
/**
 * Fires when an `upload` or `remove` operation fails.
 *
 * {% embed_file events/error/upload.component.ts %}
 * {% embed_file shared/app.component.ts %}
 * {% embed_file shared/main.ts hidden %}
 *
 * Arguments for the `error` event:
 */
var ErrorEvent = /** @class */ (function () {
    /**
     * Constructs the event arguments for the `error` event.
     *
     * @param files - The list of the files that failed to be uploaded or removed.
     * @param operation - The operation type (`upload` or `remove`).
     * @param response - The response object returned by the server.
     */
    function ErrorEvent(files, operation, response) {
        this.files = files;
        this.operation = operation;
        this.response = response;
    }
    return ErrorEvent;
}());
exports.ErrorEvent = ErrorEvent;
/**
 * Fires when an uploaded or selected file will be removed. If you cancel the event, the removal is prevented.
 *
 * {% embed_file events/remove/upload.component.ts %}
 * {% embed_file shared/app.component.ts %}
 * {% embed_file shared/main.ts hidden %}
 *
 * Arguments for the `remove` event:
 */
var RemoveEvent = /** @class */ (function (_super) {
    tslib_1.__extends(RemoveEvent, _super);
    /**
     * Constructs the event arguments for the `remove` event.
     * @param files - The list of the files that will be removed.
     * @param headers - The headers of the request.
     */
    function RemoveEvent(files, headers) {
        var _this = _super.call(this) || this;
        _this.files = files;
        _this.headers = headers;
        return _this;
    }
    return RemoveEvent;
}(PreventableEvent));
exports.RemoveEvent = RemoveEvent;
/**
 * Fires when the user selects a file or multiple files for upload. If you cancel the event, the selection is prevented.
 *
 * {% embed_file events/select/upload.component.ts preview %}
 * {% embed_file shared/app.component.ts %}
 * {% embed_file shared/main.ts hidden %}
 *
 * Arguments for the `select` event:
 */
var SelectEvent = /** @class */ (function (_super) {
    tslib_1.__extends(SelectEvent, _super);
    /**
     * Constructs the event arguments for the `select` event.
     * @param files - The list of the selected files.
     */
    function SelectEvent(files) {
        var _this = _super.call(this) || this;
        _this.files = files;
        return _this;
    }
    return SelectEvent;
}(PreventableEvent));
exports.SelectEvent = SelectEvent;
/**
 * Fires when the selected files are successfully uploaded or removed.
 *
 * {% embed_file events/success/upload.component.ts %}
 * {% embed_file shared/app.component.ts %}
 * {% embed_file shared/main.ts hidden %}
 *
 * Arguments for the `success` event:
 */
var SuccessEvent = /** @class */ (function (_super) {
    tslib_1.__extends(SuccessEvent, _super);
    /**
     * Constructs the event arguments for the `success` event.
     * @param files - The list of the files that were uploaded or removed.
     * @param operation - The operation type (`upload` or `remove`).
     * @param response - The response object returned by the server.
     */
    function SuccessEvent(files, operation, response) {
        var _this = _super.call(this) || this;
        _this.files = files;
        _this.operation = operation;
        _this.response = response;
        return _this;
    }
    return SuccessEvent;
}(PreventableEvent));
exports.SuccessEvent = SuccessEvent;
/**
 * Fires when one or more files will be uploaded. If you cancel the event, the upload is prevented.
 * You can add headers to the request.
 *
 * {% embed_file events/upload/upload.component.ts %}
 * {% embed_file shared/app.component.ts %}
 * {% embed_file shared/main.ts hidden %}
 *
 * Arguments for the `upload` event:
 */
var UploadEvent = /** @class */ (function (_super) {
    tslib_1.__extends(UploadEvent, _super);
    /**
     * Constructs the event arguments for the `upload` event.
     * @param files - The list of the files that will be uploaded.
     * @param headers - The headers of the request.
     */
    function UploadEvent(files, headers) {
        var _this = _super.call(this) || this;
        _this.files = files;
        _this.headers = headers;
        return _this;
    }
    return UploadEvent;
}(PreventableEvent));
exports.UploadEvent = UploadEvent;
/**
 * Fires while the files are being upload.
 *
 * {% embed_file events/uploadprogress/upload.component.ts %}
 * {% embed_file shared/app.component.ts %}
 * {% embed_file shared/main.ts hidden %}
 *
 * Arguments for the `uploadprogress` event:
 */
var UploadProgressEvent = /** @class */ (function () {
    /**
     * Constructs the event arguments for the `uploadprogress` event.
     * @param files - The list of files that are being uploaded.
     * @param percentComplete - The portion that has been uploaded.
     */
    function UploadProgressEvent(files, percentComplete) {
        this.files = files;
        this.percentComplete = percentComplete;
    }
    return UploadProgressEvent;
}());
exports.UploadProgressEvent = UploadProgressEvent;
