// @progress/kendo-angular-upload/dist/npm/upload-status-total.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var file_info_1 = require("./file-info");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
/**
 * @hidden
 */
var UploadStatusTotalComponent = /** @class */ (function () {
    function UploadStatusTotalComponent(localization) {
        this.localization = localization;
    }
    UploadStatusTotalComponent.prototype.ngDoCheck = function () {
        var currentlyUploading = this.fileList.hasFileWithState([file_info_1.FileState.Uploading]);
        if (this.isUploading !== currentlyUploading) {
            this.isUploading = currentlyUploading;
            this.isFailed = this.fileList.hasFileWithState([file_info_1.FileState.Failed]);
            this.statusText = this.isUploading ? this.localization.get('headerStatusUploading')
                : this.localization.get('headerStatusUploaded');
        }
    };
    /**
     * @hidden
     */
    UploadStatusTotalComponent.prototype.setStatusClasses = function () {
        return {
            "k-i-checkmark": !this.isUploading && !this.isFailed,
            "k-i-exception": !this.isUploading && this.isFailed,
            "k-i-loading": this.isUploading,
            "k-icon": true
        };
    };
    UploadStatusTotalComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'kendo-upload-status-total',
                    template: "\n        <strong\n          class=\"k-upload-status k-upload-status-total\">\n          <span [ngClass]=\"setStatusClasses()\"></span>\n          {{statusText}}\n        </strong>\n    "
                },] },
    ];
    /** @nocollapse */
    UploadStatusTotalComponent.ctorParameters = function () { return [
        { type: kendo_angular_l10n_1.LocalizationService, },
    ]; };
    UploadStatusTotalComponent.propDecorators = {
        'fileList': [{ type: core_1.Input },],
    };
    return UploadStatusTotalComponent;
}());
exports.UploadStatusTotalComponent = UploadStatusTotalComponent;
