// @progress/kendo-angular-upload/dist/es/dropzone.directive.js
import { Directive, ElementRef, HostBinding, HostListener, Input, NgZone } from '@angular/core';
import { UploadService } from './upload.service';
import util from "./util";
import validationUtil from "./validation-util";
/**
 * @hidden
 */
var DropZoneDirective = /** @class */ (function () {
    function DropZoneDirective(element, uploadService, _ngZone) {
        var _this = this;
        this.element = element;
        this.uploadService = uploadService;
        this._ngZone = _ngZone;
        this.hideIntervalDocument = null;
        this.hideIntervalElement = null;
        this.state = {
            active: "k-dropzone k-dropzone-active",
            hovered: "k-dropzone k-dropzone-active k-dropzone-hovered",
            inactive: "k-dropzone"
        };
        this.element = element;
        this.setClasses(this.state.inactive);
        if (!util.isDocumentAvailable()) {
            return;
        }
        this._ngZone.runOutsideAngular(function () {
            document.addEventListener("dragenter", _this.onDocumentDragEnterListener.bind(_this));
            document.addEventListener("dragover", _this.onDocumentDragOverListener.bind(_this));
        });
    }
    /**
     * @hidden
     */
    DropZoneDirective.prototype.onDocumentDragEnterListener = function () {
        var _this = this;
        if (this.lastDragElement) {
            if (this.calculateTimeDiff(this.lastDragElement) > 100) {
                this.setClasses(this.state.active);
            }
        }
        else {
            this.setClasses(this.state.active);
        }
        this.lastDragDocument = new Date();
        if (!this.hideIntervalDocument) {
            this.hideIntervalDocument = setInterval(function () {
                if (_this.calculateTimeDiff(_this.lastDragDocument) < 100) {
                    return;
                }
                _this.setClasses(_this.state.inactive);
                clearInterval(_this.hideIntervalDocument);
                _this.hideIntervalDocument = null;
            }, 100);
        }
        return false;
    };
    /**
     * @hidden
     */
    DropZoneDirective.prototype.onElementDragEnterListener = function () {
        var _this = this;
        this.setClasses(this.state.hovered);
        this.lastDragElement = new Date();
        if (!this.hideIntervalElement) {
            this.hideIntervalElement = setInterval(function () {
                if (_this.calculateTimeDiff(_this.lastDragElement) < 100) {
                    return;
                }
                if (_this.lastDrop) {
                    var diff = _this.calculateTimeDiff(_this.lastDrop);
                    _this.setClasses(diff > 200 ? _this.state.active : _this.state.inactive);
                }
                else {
                    _this.setClasses(_this.state.active);
                }
                clearInterval(_this.hideIntervalElement);
                _this.hideIntervalElement = null;
            }, 100);
        }
        return false;
    };
    /**
     * @hidden
     */
    DropZoneDirective.prototype.onDocumentDragOverListener = function () {
        this.lastDragDocument = new Date();
        return false;
    };
    /**
     * @hidden
     */
    DropZoneDirective.prototype.onElementDragOverListener = function () {
        this.lastDragElement = new Date();
        return false;
    };
    /**
     * @hidden
     */
    DropZoneDirective.prototype.onDropListener = function (event) {
        this.lastDrop = new Date();
        var droppedFiles = event.dataTransfer.files;
        if (droppedFiles.length > 0 && !this.disabled) {
            var files = util.getAllFileInfo(droppedFiles);
            files = util.assignGuidToFiles(files, !this.async.batch);
            if (!this.multiple) {
                files.splice(1, files.length - 1);
                this.uploadService.clearFiles();
            }
            validationUtil.validateFiles(files, this.restrictions);
            this.uploadService.addFiles(files, this.async);
        }
        return false;
    };
    DropZoneDirective.prototype.calculateTimeDiff = function (prevEvent) {
        return new Date().getTime() - prevEvent.getTime();
    };
    DropZoneDirective.prototype.setClasses = function (classes) {
        if (!this.disabled) {
            this.element.nativeElement.className = classes;
        }
    };
    Object.defineProperty(DropZoneDirective.prototype, "initialClassName", {
        /**
         * @hidden
         */
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    DropZoneDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[kendoUploadDropzone]'
                },] },
    ];
    /** @nocollapse */
    DropZoneDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: UploadService, },
        { type: NgZone, },
    ]; };
    DropZoneDirective.propDecorators = {
        'async': [{ type: Input },],
        'disabled': [{ type: Input },],
        'multiple': [{ type: Input },],
        'restrictions': [{ type: Input },],
        'onElementDragEnterListener': [{ type: HostListener, args: ['dragenter',] },],
        'onElementDragOverListener': [{ type: HostListener, args: ['dragover',] },],
        'onDropListener': [{ type: HostListener, args: ['drop', ['$event'],] },],
        'initialClassName': [{ type: HostBinding, args: ['class.k-dropzone',] },],
    };
    return DropZoneDirective;
}());
export { DropZoneDirective };
