// @progress/kendo-angular-upload/dist/es/upload-status-total.component.js
import { Component, Input } from '@angular/core';
import { FileState } from './file-info';
import { LocalizationService } from '@progress/kendo-angular-l10n';
/**
 * @hidden
 */
var UploadStatusTotalComponent = /** @class */ (function () {
    function UploadStatusTotalComponent(localization) {
        this.localization = localization;
    }
    UploadStatusTotalComponent.prototype.ngDoCheck = function () {
        var currentlyUploading = this.fileList.hasFileWithState([FileState.Uploading]);
        if (this.isUploading !== currentlyUploading) {
            this.isUploading = currentlyUploading;
            this.isFailed = this.fileList.hasFileWithState([FileState.Failed]);
            this.statusText = this.isUploading ? this.localization.get('headerStatusUploading')
                : this.localization.get('headerStatusUploaded');
        }
    };
    /**
     * @hidden
     */
    UploadStatusTotalComponent.prototype.setStatusClasses = function () {
        return {
            "k-i-checkmark": !this.isUploading && !this.isFailed,
            "k-i-exception": !this.isUploading && this.isFailed,
            "k-i-loading": this.isUploading,
            "k-icon": true
        };
    };
    UploadStatusTotalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'kendo-upload-status-total',
                    template: "\n        <strong\n          class=\"k-upload-status k-upload-status-total\">\n          <span [ngClass]=\"setStatusClasses()\"></span>\n          {{statusText}}\n        </strong>\n    "
                },] },
    ];
    /** @nocollapse */
    UploadStatusTotalComponent.ctorParameters = function () { return [
        { type: LocalizationService, },
    ]; };
    UploadStatusTotalComponent.propDecorators = {
        'fileList': [{ type: Input },],
    };
    return UploadStatusTotalComponent;
}());
export { UploadStatusTotalComponent };
