// @progress/kendo-angular-sortable/dist/es/sortable.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SortableComponent } from './sortable.component';
import { SortableService } from "./sortable.service";
var COMPONENT_DIRECTIVES = [SortableComponent];
/**
 * Represents the [`NgModule`](https://angular.io/docs/ts/latest/guide/ngmodule.html)
 * definition for the Sortable component.
 *
 * @example
 *
 * ```ts-no-run
 * import { NgModule } from '@angular/core';
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * // Import the Sortable module
 * import { SortableModule } from '@progress/kendo-angular-sortable';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, SortableModule],
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
var SortableModule = /** @class */ (function () {
    function SortableModule() {
    }
    SortableModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [COMPONENT_DIRECTIVES],
                    exports: [COMPONENT_DIRECTIVES],
                    imports: [CommonModule],
                    providers: [SortableService]
                },] },
    ];
    /** @nocollapse */
    SortableModule.ctorParameters = function () { return []; };
    return SortableModule;
}());
export { SortableModule };
