// @progress/kendo-angular-treeview/dist/npm/navigation/navigation.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var navigation_model_1 = require("./navigation-model");
var utils_1 = require("../utils");
var keys_1 = require("./keys");
var Subject_1 = require("rxjs/Subject");
var sibling = function (item, direction) { return item[direction] || item; };
/**
 * @hidden
 */
var NavigationService = /** @class */ (function () {
    function NavigationService() {
        var _this = this;
        this.expands = new Subject_1.Subject();
        this.moves = new Subject_1.Subject();
        this.checks = new Subject_1.Subject();
        this.selects = new Subject_1.Subject();
        this.navigable = true;
        this.actions = (_a = {},
            _a[keys_1.Keys.up] = function () { return _this.activate(sibling(_this.focusableItem, 'prev')); },
            _a[keys_1.Keys.down] = function () { return _this.activate(sibling(_this.focusableItem, 'next')); },
            _a[keys_1.Keys.left] = function () { return (_this.expand({ expand: false, intercept: _this.moveToParent.bind(_this) })); },
            _a[keys_1.Keys.right] = function () { return (_this.expand({ expand: true, intercept: _this.moveToChild.bind(_this) })); },
            _a[keys_1.Keys.home] = function () { return _this.activate(_this.model.firstNode()); },
            _a[keys_1.Keys.end] = function () { return _this.activate(_this.model.lastNode()); },
            _a[keys_1.Keys.enter] = function () { return _this.navigable && _this.selectIndex(utils_1.nodeIndex(_this.activeItem)); },
            _a[keys_1.Keys.space] = function () { return _this.navigable && _this.checkIndex(utils_1.nodeIndex(_this.activeItem)); },
            _a);
        this.isFocused = false;
        this._model = new navigation_model_1.NavigationModel();
        var _a;
    }
    Object.defineProperty(NavigationService.prototype, "model", {
        get: function () {
            return this._model;
        },
        set: function (model) {
            this._model = model;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavigationService.prototype, "activeIndex", {
        get: function () {
            return utils_1.nodeIndex(this.activeItem) || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavigationService.prototype, "focusableItem", {
        get: function () {
            return this.activeItem || this.model.firstNode();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavigationService.prototype, "isActiveExpanded", {
        get: function () {
            return this.activeItem && this.activeItem.children.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    NavigationService.prototype.activate = function (item) {
        if (!this.navigable || !item || this.isActive(utils_1.nodeIndex(item))) {
            return;
        }
        this.isFocused = true;
        this.activeItem = item || this.activeItem;
        this.notifyMove();
    };
    NavigationService.prototype.activateParent = function (index) {
        this.activate(this.model.findParent(index));
    };
    NavigationService.prototype.activateIndex = function (index) {
        if (!index) {
            return;
        }
        this.activate(this.model.findNode(index));
    };
    NavigationService.prototype.activateClosest = function (index) {
        if (!index || utils_1.nodeIndex(this.focusableItem) !== index) {
            return;
        }
        this.activeItem = this.model.closestNode(index);
        this.notifyMove();
    };
    NavigationService.prototype.activateFocusable = function () {
        if (this.activeItem) {
            return;
        }
        this.activeItem = this.model.firstNode();
        this.notifyMove();
    };
    NavigationService.prototype.deactivate = function () {
        if (!this.navigable || !this.isFocused) {
            return;
        }
        this.isFocused = false;
        this.notifyMove();
    };
    NavigationService.prototype.checkIndex = function (index) {
        this.checks.next(index);
    };
    NavigationService.prototype.selectIndex = function (index) {
        this.selects.next(index);
    };
    NavigationService.prototype.isActive = function (index) {
        if (!index) {
            return false;
        }
        return this.isFocused && this.activeIndex === index;
    };
    NavigationService.prototype.isFocusable = function (index) {
        return utils_1.nodeIndex(this.focusableItem) === index;
    };
    NavigationService.prototype.registerItem = function (index) {
        this.model.registerItem(index);
    };
    NavigationService.prototype.unregisterItem = function (index) {
        if (this.isActive(index)) {
            this.activateParent(index);
        }
        this.model.unregisterItem(index);
    };
    NavigationService.prototype.move = function (e) {
        if (!this.navigable) {
            return;
        }
        var moveAction = this.actions[e.keyCode];
        if (!moveAction) {
            return;
        }
        moveAction();
        e.preventDefault();
    };
    NavigationService.prototype.expand = function (_a) {
        var expand = _a.expand, intercept = _a.intercept;
        var index = utils_1.nodeIndex(this.activeItem);
        if (!index || intercept(index)) {
            return;
        }
        this.notifyExpand(expand);
    };
    NavigationService.prototype.moveToParent = function () {
        if (this.isActiveExpanded) {
            return false;
        }
        this.activate(this.model.findParent(utils_1.nodeIndex(this.activeItem)));
        return true;
    };
    NavigationService.prototype.moveToChild = function () {
        if (!this.isActiveExpanded) {
            return false;
        }
        this.activate(this.model.findChild(utils_1.nodeIndex(this.activeItem)));
        return true;
    };
    NavigationService.prototype.notifyExpand = function (expand) {
        this.expands.next(this.navigationState(expand));
    };
    NavigationService.prototype.notifyMove = function () {
        this.moves.next(this.navigationState());
    };
    NavigationService.prototype.navigationState = function (expand) {
        if (expand === void 0) { expand = false; }
        return ({ expand: expand, index: utils_1.nodeIndex(this.activeItem), isFocused: this.isFocused });
    };
    NavigationService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    NavigationService.ctorParameters = function () { return []; };
    return NavigationService;
}());
exports.NavigationService = NavigationService;
