// @progress/kendo-angular-treeview/dist/npm/treeview.component.js
"use strict";
/* tslint:disable:member-ordering */
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var node_template_directive_1 = require("./node-template.directive");
var focus_state_service_1 = require("./focus-state.service");
var expand_state_service_1 = require("./expand-state.service");
var index_builder_service_1 = require("./index-builder.service");
var loading_notification_service_1 = require("./loading-notification.service");
var node_children_service_1 = require("./node-children.service");
var navigation_service_1 = require("./navigation/navigation.service");
var selection_service_1 = require("./selection/selection.service");
var treeview_lookup_service_1 = require("./treeview-lookup.service");
var utils_1 = require("./utils");
var default_callbacks_1 = require("./default-callbacks");
var Subscription_1 = require("rxjs/Subscription");
var BehaviorSubject_1 = require("rxjs/BehaviorSubject");
var auditTime_1 = require("rxjs/operators/auditTime");
var of_1 = require("rxjs/observable/of");
var isExpandIcon = function (target) { return (/k-i-collapse|k-i-expand/.test(target.className)); };
var providers = [
    focus_state_service_1.FocusStateService,
    expand_state_service_1.ExpandStateService,
    index_builder_service_1.IndexBuilderService,
    treeview_lookup_service_1.TreeViewLookupService,
    loading_notification_service_1.LoadingNotificationService,
    node_children_service_1.NodeChildrenService,
    navigation_service_1.NavigationService,
    selection_service_1.SelectionService
];
/**
 * Represents the Kendo UI TreeView component for Angular.
 *
 * @example
 * {% meta height:350 %}
 * ```ts-preview
 * import { Component } from '@angular/core';
 *
 *  _@Component({
 *      selector: 'my-app',
 *      template: `
 *      <kendo-treeview
 *          [nodes]="data"
 *          textField="text"
 *          kendoTreeViewCheckable
 *          kendoTreeViewExpandable
 *          kendoTreeViewSelectable
 *
 *          kendoTreeViewHierarchyBinding
 *          childrenField="items">
 *      </kendo-treeview>
 *  `
 *  })
 *  export class AppComponent {
 *      public data: any[] = [
 *          {
 *              text: "Furniture", items: [
 *                  { text: "Tables & Chairs" },
 *                  { text: "Sofas" },
 *                  { text: "Occasional Furniture" }
 *              ]
 *          },
 *          {
 *              text: "Decor", items: [
 *                  { text: "Bed Linen" },
 *                  { text: "Curtains & Blinds" },
 *                  { text: "Carpets" }
 *              ]
 *          }
 *      ];
 *  }
 * ```
 * {% endmeta %}
 */
var TreeViewComponent = /** @class */ (function () {
    function TreeViewComponent(focusService, expandService, navigationService, nodeChildrenService, selectionService, treeViewLookupService, element) {
        this.focusService = focusService;
        this.expandService = expandService;
        this.navigationService = navigationService;
        this.nodeChildrenService = nodeChildrenService;
        this.selectionService = selectionService;
        this.treeViewLookupService = treeViewLookupService;
        this.element = element;
        this.classNames = true;
        /**
         * Fires when the children of the expanded node are loaded.
         */
        this.childrenLoaded = new core_1.EventEmitter();
        /**
         * Fires when the user blurs the component.
         */
        this.onBlur = new core_1.EventEmitter(); //tslint:disable-line:no-output-rename
        /**
         * Fires when the user focuses the component.
         */
        this.onFocus = new core_1.EventEmitter(); //tslint:disable-line:no-output-rename
        /**
         * Fires when the user expands a TreeView node.
         */
        this.expand = new core_1.EventEmitter();
        /**
         * Fires when the user collapses a TreeView node.
         */
        this.collapse = new core_1.EventEmitter();
        /**
         * Fires when the user selects a TreeView node checkbox.
         */
        this.checkedChange = new core_1.EventEmitter();
        /**
         * Fires when the user selects a TreeView node.
         */
        this.selectionChange = new core_1.EventEmitter();
        /**
         * A function which determines if a specific node is disabled.
         */
        this.isDisabled = default_callbacks_1.isDisabled;
        /**
         * Determines whether the TreeView keyboard navigable is enabled.
         */
        this.navigable = true;
        /**
         * A function which provides the child nodes for a given parent node.
         */
        this.children = function () { return of_1.of([]); };
        this.checkboxes = false;
        this.expandIcons = false;
        this.isActive = false;
        this.data = new BehaviorSubject_1.BehaviorSubject([]);
        this._animate = true;
        this.resolvedPromise = Promise.resolve(null);
        this.subscriptions = new Subscription_1.Subscription(function () { });
    }
    Object.defineProperty(TreeViewComponent.prototype, "role", {
        get: function () { return 'tree'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeViewComponent.prototype, "animate", {
        get: function () {
            return !this._animate;
        },
        /**
         * Determines whether the content animation is enabled.
         */
        set: function (value) {
            this._animate = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeViewComponent.prototype, "nodes", {
        /**
         * The nodes which will be displayed by the TreeView.
         */
        set: function (value) {
            this.data.next(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeViewComponent.prototype, "hasChildren", {
        /**
         * A function which determines if a specific node has child nodes.
         */
        get: function () {
            return this._hasChildren || default_callbacks_1.hasChildren;
        },
        set: function (callback) {
            this._hasChildren = callback;
            this.expandIcons = Boolean(this._isExpanded && this._hasChildren);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeViewComponent.prototype, "isChecked", {
        /**
         * A function which determines if a specific node is selected.
         */
        get: function () {
            return this._isChecked || default_callbacks_1.isChecked;
        },
        set: function (callback) {
            this._isChecked = callback;
            this.checkboxes = Boolean(this._isChecked);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeViewComponent.prototype, "isExpanded", {
        /**
         * A function which determines if a specific node is expanded.
         */
        get: function () {
            return this._isExpanded || default_callbacks_1.isExpanded;
        },
        set: function (callback) {
            this._isExpanded = callback;
            this.expandIcons = Boolean(this._isExpanded && this._hasChildren);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeViewComponent.prototype, "isSelected", {
        /**
         * A function which determines if a specific node is selected.
         */
        get: function () {
            return this._isSelected || default_callbacks_1.isSelected;
        },
        set: function (callback) {
            this._isSelected = callback;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @hidden
     */
    TreeViewComponent.prototype.handleKeydown = function (args) {
        if (!this.isActive) {
            return;
        }
        this.navigationService.move(args);
    };
    /**
     * @hidden
     */
    TreeViewComponent.prototype.handleClick = function (_a) {
        var _this = this;
        var target = _a.target;
        var index = utils_1.nodeId(utils_1.closestNode(target));
        if (!index || this.focusService.hasActiveContent()) {
            return;
        }
        this.focusin(index)
            .then(function () {
            if (!_this.isActive || isExpandIcon(target)) {
                return;
            }
            _this.navigationService.selectIndex(index);
        });
    };
    TreeViewComponent.prototype.ngOnChanges = function (_) {
        this.navigationService.navigable = Boolean(this.navigable);
    };
    TreeViewComponent.prototype.ngOnDestroy = function () {
        this.subscriptions.unsubscribe();
    };
    TreeViewComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.subscriptions.add(this.nodeChildrenService
            .changes
            .subscribe(function (x) { return _this.childrenLoaded.emit(x); }));
        this.subscriptions.add(this.focusService.changes
            .pipe(auditTime_1.auditTime(1))
            .subscribe(function (_a) {
            var focused = _a.focused, index = _a.index;
            return focused
                ? _this.focusin(index)
                : _this.focusout(index);
        }));
        this.subscriptions.add(this.expandService.changes
            .subscribe(function (_a) {
            var index = _a.index, dataItem = _a.dataItem, expand = _a.expand;
            return expand
                ? _this.expand.emit({ index: index, dataItem: dataItem })
                : _this.collapse.emit({ index: index, dataItem: dataItem });
        }));
        this.subscriptions.add(this.navigationService.checks
            .subscribe(function (x) {
            return _this.checkedChange.emit(_this.treeViewLookupService.itemLookup(x));
        }));
        this.subscriptions.add(this.selectionService.changes
            .subscribe(function (x) { return _this.selectionChange.emit(x); }));
    };
    /**
     * Blurs the focused TreeView item.
     */
    TreeViewComponent.prototype.blur = function () {
        if (!utils_1.isDocumentAvailable()) {
            return;
        }
        var target = utils_1.focusableNode(this.element);
        if (document.activeElement === target) {
            target.blur();
        }
    };
    /**
     * Focuses the first focusable item in the TreeView component if no hierarchical index is provided.
     *
     * @example
     * ```ts
     * import { Component } from '@angular/core';
     *
     *  @Component({
     *      selector: 'my-app',
     *      template: `
     *      <button (click)="treeview.focus('1')">Focuses the second node</button>
     *      <kendo-treeview
     *          #treeview
     *          [nodes]="data"
     *          textField="text"
     *      >
     *      </kendo-treeview>
     *  `
     *  })
     *  export class AppComponent {
     *      public data: any[] = [
     *          { text: "Furniture" },
     *          { text: "Decor" }
     *      ];
     *  }
     * ```
     */
    TreeViewComponent.prototype.focus = function (index) {
        this.navigationService.activateIndex(index);
        var target = utils_1.focusableNode(this.element);
        if (target) {
            target.focus();
        }
    };
    TreeViewComponent.prototype.focusin = function (index) {
        var _this = this;
        if (!index) {
            return;
        }
        this.navigationService.activateIndex(index);
        return this.resolvedPromise.then(function () {
            if (!_this.isActive && _this.navigationService.isActive(index)) {
                _this.isActive = true;
                _this.onFocus.emit();
            }
        });
    };
    TreeViewComponent.prototype.focusout = function (index) {
        var _this = this;
        return this.resolvedPromise.then(function () {
            if (_this.isActive && _this.navigationService.isActive(index)) {
                _this.navigationService.deactivate();
                _this.isActive = false;
                _this.onBlur.emit();
            }
        });
    };
    TreeViewComponent.decorators = [
        { type: core_1.Component, args: [{
                    changeDetection: core_1.ChangeDetectionStrategy.Default,
                    exportAs: 'kendoTreeView',
                    providers: providers,
                    selector: 'kendo-treeview',
                    template: "\n    <ul class=\"k-treeview-lines\"\n      kendoTreeViewGroup\n      role=\"group\"\n      [checkboxes]=\"checkboxes\"\n      [expandIcons]=\"expandIcons\"\n      [children]=\"children\"\n      [hasChildren]=\"hasChildren\"\n      [isChecked]=\"isChecked\"\n      [isDisabled]=\"isDisabled\"\n      [isExpanded]=\"isExpanded\"\n      [isSelected]=\"isSelected\"\n      [nodeTemplateRef]=\"nodeTemplate?.templateRef\"\n      [nodes]=\"data\"\n      [textField]=\"textField\"\n      >\n    </ul>\n  "
                },] },
    ];
    /** @nocollapse */
    TreeViewComponent.ctorParameters = function () { return [
        { type: focus_state_service_1.FocusStateService, },
        { type: expand_state_service_1.ExpandStateService, },
        { type: navigation_service_1.NavigationService, },
        { type: node_children_service_1.NodeChildrenService, },
        { type: selection_service_1.SelectionService, },
        { type: treeview_lookup_service_1.TreeViewLookupService, },
        { type: core_1.ElementRef, },
    ]; };
    TreeViewComponent.propDecorators = {
        'classNames': [{ type: core_1.HostBinding, args: ["class.k-widget",] }, { type: core_1.HostBinding, args: ["class.k-treeview",] },],
        'role': [{ type: core_1.HostBinding, args: ["attr.role",] },],
        'animate': [{ type: core_1.Input }, { type: core_1.HostBinding, args: ['@.disabled',] },],
        'childrenLoaded': [{ type: core_1.Output },],
        'onBlur': [{ type: core_1.Output, args: ['blur',] },],
        'onFocus': [{ type: core_1.Output, args: ['focus',] },],
        'expand': [{ type: core_1.Output },],
        'collapse': [{ type: core_1.Output },],
        'checkedChange': [{ type: core_1.Output },],
        'selectionChange': [{ type: core_1.Output },],
        'nodeTemplate': [{ type: core_1.ContentChild, args: [node_template_directive_1.NodeTemplateDirective,] },],
        'nodes': [{ type: core_1.Input },],
        'textField': [{ type: core_1.Input },],
        'hasChildren': [{ type: core_1.Input },],
        'isChecked': [{ type: core_1.Input },],
        'isDisabled': [{ type: core_1.Input },],
        'isExpanded': [{ type: core_1.Input },],
        'isSelected': [{ type: core_1.Input },],
        'navigable': [{ type: core_1.Input },],
        'children': [{ type: core_1.Input },],
        'handleKeydown': [{ type: core_1.HostListener, args: ['keydown', ['$event'],] },],
        'handleClick': [{ type: core_1.HostListener, args: ['click', ['$event'],] },],
    };
    return TreeViewComponent;
}());
exports.TreeViewComponent = TreeViewComponent;
