// @progress/kendo-angular-grid/dist/npm/navigation/navigation.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var from_1 = require("rxjs/observable/from");
var filter_1 = require("rxjs/operators/filter");
var switchMap_1 = require("rxjs/operators/switchMap");
var switchMapTo_1 = require("rxjs/operators/switchMapTo");
var take_1 = require("rxjs/operators/take");
var takeWhile_1 = require("rxjs/operators/takeWhile");
var dom_events_service_1 = require("../common/dom-events.service");
var groups_service_1 = require("../grouping/groups.service");
var pager_context_service_1 = require("../pager/pager-context.service");
var dom_queries_1 = require("../rendering/common/dom-queries");
var cell_keydown_event_1 = require("./cell-keydown-event");
var focus_root_1 = require("./focus-root");
var navigation_cursor_1 = require("./navigation-cursor");
var navigation_model_1 = require("./navigation-model");
var scroll_request_service_1 = require("../scrolling/scroll-request.service");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var edit_service_1 = require("../editing/edit.service");
var targetCell = function (target) {
    var table = dom_queries_1.closest(target, dom_queries_1.matchesNodeName('table'));
    var heading = dom_queries_1.closestInScope(target, dom_queries_1.matchesNodeName('th'), table);
    var cell = dom_queries_1.closestInScope(target, dom_queries_1.matchesNodeName('td'), table) || heading;
    var row = dom_queries_1.closestInScope(cell, dom_queries_1.matchesNodeName('tr'), table);
    if (cell && row) {
        var rowIndex = row.getAttribute('aria-rowindex');
        rowIndex = rowIndex ? parseInt(rowIndex, 10) - 1 : null;
        var colIndex = cell.getAttribute('aria-colindex');
        colIndex = colIndex ? parseInt(colIndex, 10) - 1 : null;
        if (rowIndex !== null && colIndex !== null) {
            return { colIndex: colIndex, rowIndex: rowIndex, element: cell };
        }
    }
};
var lastCell = function (row) {
    var cells = row.cells;
    return cells[cells.length - 1];
};
var isArrowKey = function (keyCode) {
    return keyCode === 37 /* left */ || keyCode === 39 /* right */ ||
        keyCode === 38 /* up */ || keyCode === 40 /* down */;
};
var isNavigationKey = function (keyCode) {
    return isArrowKey(keyCode) ||
        keyCode === 33 /* page_up */ || keyCode === 34 /* page_down */ ||
        keyCode === 36 /* home */ || keyCode === 35 /* end */;
};
var isInput = dom_queries_1.matchesNodeName('input');
var isTextInput = function (element) {
    return element && isInput(element) && element.type.toLowerCase() === 'text';
};
/**
 * @hidden
 */
var NavigationViewport = /** @class */ (function () {
    function NavigationViewport(firstItemIndex, lastItemIndex) {
        this.firstItemIndex = firstItemIndex;
        this.lastItemIndex = lastItemIndex;
    }
    NavigationViewport.prototype.containsRow = function (dataRowIndex) {
        var headerRow = dataRowIndex < 0;
        return headerRow || (dataRowIndex >= this.firstItemIndex && dataRowIndex < this.lastItemIndex);
    };
    return NavigationViewport;
}());
exports.NavigationViewport = NavigationViewport;
/**
 * @hidden
 */
var NavigationService = /** @class */ (function () {
    function NavigationService(zone, domEvents, pagerContextService, scrollRequestService, groupsService, focusRoot, editService, cd, rtl) {
        this.zone = zone;
        this.domEvents = domEvents;
        this.pagerContextService = pagerContextService;
        this.scrollRequestService = scrollRequestService;
        this.groupsService = groupsService;
        this.focusRoot = focusRoot;
        this.editService = editService;
        this.cd = cd;
        this.rtl = rtl;
        this.cellKeydown = new core_1.EventEmitter();
        this.activeRow = 0;
        this.alive = false;
        this.mode = 0 /* Standby */;
        this.model = new navigation_model_1.NavigationModel();
        this.cursor = new navigation_cursor_1.NavigationCursor(this.model);
        this.changes = this.cursor.changes;
    }
    Object.defineProperty(NavigationService.prototype, "metadata", {
        get: function () {
            return this.meta;
        },
        set: function (value) {
            this.meta = value;
            this.cursor.metadata = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavigationService.prototype, "enabled", {
        get: function () {
            return this.alive;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavigationService.prototype, "activeDataRow", {
        get: function () {
            return this.activeRow - this.meta.headerRows;
        },
        enumerable: true,
        configurable: true
    });
    NavigationService.prototype.init = function (meta) {
        var _this = this;
        this.alive = true;
        var isAlive = function () { return _this.alive; };
        this.metadata = meta;
        var onStable = function () { return _this.zone.isStable ?
            from_1.from([true]) :
            _this.zone.onStable.pipe(take_1.take(1)); };
        this.cursor.changes
            .pipe(takeWhile_1.takeWhile(isAlive))
            .subscribe(function (args) {
            _this.activeRow = args.rowIndex;
            if (_this.meta && _this.meta.isVirtual &&
                _this.viewport && !_this.viewport.containsRow(_this.activeDataRow)) {
                _this.scrollRequestService.scrollTo(_this.activeDataRow);
            }
        });
        this.domEvents.focus.pipe(takeWhile_1.takeWhile(isAlive), switchMap_1.switchMap(onStable, function (args) { return args; }))
            .subscribe(function (args) { return _this.navigateTo(args.target); });
        this.domEvents.focusOut.pipe(takeWhile_1.takeWhile(isAlive), filter_1.filter(function () { return _this.mode !== 0 /* Standby */; }), switchMap_1.switchMap(onStable, function (args) { return args; }))
            .subscribe(function (args) {
            if (dom_queries_1.isVisible(args.target)) {
                _this.mode = 0 /* Standby */;
            }
            else {
                // Focused target is no longer visible,
                // reset to cursor mode and recapture focus.
                _this.mode = 1 /* Cursor */;
            }
            _this.deactivateElements();
            _this.cursor.announce();
        });
        this.domEvents.windowBlur.pipe(takeWhile_1.takeWhile(isAlive), filter_1.filter(function () { return _this.mode !== 0 /* Standby */; }))
            .subscribe(function () {
            _this.mode = 0 /* Standby */;
            _this.deactivateElements();
            _this.cursor.announce();
        });
        // Closing the editor will not always trigger focusout in Firefox.
        // To get around this, we ensure that the cell is closed after editing.
        this.editService.changes.pipe(takeWhile_1.takeWhile(isAlive), filter_1.filter(function (e) { return e.action !== 'edit' && _this.mode === 2 /* Content */; }), switchMap_1.switchMap(onStable, function (args) { return args; }))
            .subscribe(function () { return _this.leaveCell(); });
        this.pagerContextService.pageChange
            .pipe(takeWhile_1.takeWhile(isAlive))
            .subscribe(function () { return _this.cursor.reset(0, 0); });
        this.domEvents.keydown
            .pipe(takeWhile_1.takeWhile(isAlive))
            .subscribe(function (args) {
            if (_this.mode === 1 /* Cursor */) {
                _this.onCursorKeydown(args);
            }
            else if (_this.mode === 2 /* Content */) {
                _this.onContentKeydown(args);
            }
        });
        this.domEvents.keydown.pipe(takeWhile_1.takeWhile(isAlive), filter_1.filter(function (args) { return args.keyCode === 9 /* tab */ && _this.mode === 2 /* Content */; }), switchMapTo_1.switchMapTo(this.domEvents.focusOut.pipe(take_1.take(1))))
            .subscribe(function () {
            // Tabbed out of the last focusable content element
            // reset to cursor mode and recapture focus.
            _this.leaveCell();
            _this.cursor.reset();
        });
        this.deactivateElements();
    };
    NavigationService.prototype.ngOnDestroy = function () {
        this.alive = false;
    };
    NavigationService.prototype.registerCell = function (cell) {
        this.model.registerCell(cell);
    };
    NavigationService.prototype.registerRow = function (row) {
        this.model.registerRow(row);
    };
    NavigationService.prototype.unregisterRow = function (row) {
        this.model.unregisterRow(row);
    };
    NavigationService.prototype.isCellFocusable = function (cell) {
        return this.alive && this.mode !== 2 /* Content */ && this.cursor.isActive(cell.logicalRowIndex, cell.logicalColIndex);
    };
    NavigationService.prototype.isCellFocused = function (cell) {
        return this.mode === 1 /* Cursor */ && this.isCellFocusable(cell);
    };
    NavigationService.prototype.navigateTo = function (el) {
        var cell = targetCell(el);
        if (!cell) {
            return;
        }
        var oldMode = this.mode;
        var focusInCell = dom_queries_1.contains(cell.element, document.activeElement);
        var focusInActiveRowContent = this.mode === 2 /* Content */ && this.activeRow === cell.rowIndex && el !== cell.element;
        if (focusInCell) {
            this.mode = 2 /* Content */;
            this.cursor.reset(cell.rowIndex, cell.colIndex);
            this.activateRow();
        }
        else if (!focusInActiveRowContent) {
            this.mode = 1 /* Cursor */;
            this.deactivateElements();
            var alreadyActive = this.cursor.isActive(cell.rowIndex, cell.colIndex);
            var isCursor = oldMode === 1 /* Cursor */ && alreadyActive;
            if (!isCursor) {
                this.cursor.reset(cell.rowIndex, cell.colIndex);
            }
        }
    };
    NavigationService.prototype.tryFocus = function (el) {
        this.activateElements();
        var focusable = dom_queries_1.findFocusableChild(el);
        if (focusable) {
            var cell = targetCell(focusable);
            if (cell) {
                this.cursor.reset(cell.rowIndex, cell.colIndex);
                this.deactivateElements();
                this.enterCell();
            }
            focusable.focus();
        }
        else {
            this.deactivateElements();
        }
        return !!focusable;
    };
    NavigationService.prototype.needsViewport = function () {
        return this.meta && this.meta.isVirtual;
    };
    NavigationService.prototype.setViewport = function (firstItemIndex, lastItemIndex) {
        this.viewport = new NavigationViewport(firstItemIndex, lastItemIndex);
        if (this.meta && this.meta.isVirtual && lastItemIndex > 0) {
            var dataRowIndex = Math.max(0, this.activeRow - this.meta.headerRows);
            var ahead = firstItemIndex - dataRowIndex;
            var behind = dataRowIndex - lastItemIndex;
            if (ahead > 0) {
                this.cursor.reset(firstItemIndex + this.meta.headerRows + 1);
            }
            else if (behind > 0) {
                this.cursor.reset(lastItemIndex + this.meta.headerRows - 1);
            }
        }
    };
    NavigationService.prototype.enterCell = function () {
        var cell = this.cursor.cell;
        if (!cell) {
            return;
        }
        var group = cell.focusGroup;
        var focusable = group && group.canFocus();
        this.mode = focusable ? 2 /* Content */ : 1 /* Cursor */;
        this.cursor.announce();
        if (focusable) {
            this.activateRow();
            group.focus();
        }
    };
    NavigationService.prototype.leaveCell = function () {
        var cell = this.cursor.cell;
        if (!cell) {
            return;
        }
        var group = cell.focusGroup;
        var focusable = group && group.canFocus();
        if (!focusable) {
            this.deactivateElements();
        }
        this.mode = 1 /* Cursor */;
        this.cursor.announce();
    };
    NavigationService.prototype.activateElements = function () {
        this.focusRoot.activate();
    };
    NavigationService.prototype.deactivateElements = function () {
        this.focusRoot.deactivate();
    };
    NavigationService.prototype.activateRow = function () {
        this.cursor.row.cells
            .forEach(function (cell) { return cell.focusGroup && cell.focusGroup.activate(); });
    };
    NavigationService.prototype.onCursorKeydown = function (args) {
        var _this = this;
        var preventDefault = false;
        var modifier = args.ctrlKey || args.metaKey;
        var step = modifier ? 5 : 1;
        if (!this.onCellKeydown(args)) {
            return;
        }
        switch (args.keyCode) {
            case 40 /* down */:
                preventDefault = this.cursor.moveDown(step);
                break;
            case 38 /* up */:
                preventDefault = this.cursor.moveUp(step);
                break;
            case 39 /* right */:
                preventDefault = this.rtl ? this.cursor.moveLeft() : this.cursor.moveRight();
                break;
            case 37 /* left */:
                preventDefault = this.rtl ? this.cursor.moveRight() : this.cursor.moveLeft();
                break;
            case 34 /* page_down */:
                if (this.metadata.isVirtual && this.viewport) {
                    this.cursor.reset(this.viewport.lastItemIndex + this.meta.headerRows);
                    preventDefault = true;
                }
                else if (this.metadata.hasPager) {
                    this.zone.run(function () { return _this.pagerContextService.nextPage(); });
                    preventDefault = true;
                }
                break;
            case 33 /* page_up */:
                if (this.metadata.isVirtual && this.viewport) {
                    var viewportSize = this.viewport.lastItemIndex - this.viewport.firstItemIndex;
                    var nextItemIndex = Math.max(0, this.viewport.firstItemIndex - viewportSize);
                    this.cursor.reset(nextItemIndex + this.meta.headerRows);
                    preventDefault = true;
                }
                else if (this.metadata.hasPager) {
                    this.zone.run(function () { return _this.pagerContextService.prevPage(); });
                    preventDefault = true;
                }
                break;
            case 36 /* home */:
                if (modifier) {
                    if (this.meta.isVirtual) {
                        this.cursor.reset(this.meta.headerRows, 0);
                    }
                    else {
                        this.cursor.reset(this.model.firstRow.index, 0);
                    }
                }
                else {
                    var row = this.cursor.row;
                    var cells = row.cells;
                    var firstCell = cells[0];
                    this.cursor.reset(row.index, firstCell.colIndex);
                }
                preventDefault = true;
                break;
            case 35 /* end */:
                if (modifier) {
                    var colIndex = lastCell(this.model.lastRow).colIndex;
                    if (this.meta.isVirtual) {
                        var lastRowIndex = this.meta.dataRows - 1;
                        this.cursor.reset(this.meta.headerRows + lastRowIndex, colIndex);
                    }
                    else {
                        this.cursor.reset(this.model.lastRow.index, colIndex);
                    }
                }
                else {
                    var row = this.cursor.row;
                    var colIndex = lastCell(row).colIndex;
                    this.cursor.reset(row.index, colIndex);
                }
                preventDefault = true;
                break;
            case 13 /* enter */:
            case 113 /* f2 */:
                var groupItem_1 = this.cursor.row.groupItem;
                if (groupItem_1) {
                    this.zone.run(function () {
                        return _this.groupsService.toggleRow(groupItem_1.index, groupItem_1.data);
                    });
                }
                else {
                    this.enterCell();
                    if (!this.cursor.cell.focusGroup.isNavigable()) {
                        preventDefault = true;
                    }
                }
                break;
            default:
                if (!args.ctrlKey && !args.altKey) {
                    var isAlphaNum = args.keyCode >= 48 && args.keyCode <= 90;
                    var isNumpad = args.keyCode >= 96 && args.keyCode <= 111;
                    if (isAlphaNum || isNumpad) {
                        this.enterCell();
                    }
                }
        }
        if (preventDefault) {
            args.preventDefault();
        }
    };
    NavigationService.prototype.onContentKeydown = function (args) {
        if (!this.onCellKeydown(args)) {
            return;
        }
        var confirm = !args.defaultPrevented && args.keyCode === 13 /* enter */ && isTextInput(args.srcElement);
        if (args.keyCode === 27 /* esc */ || args.keyCode === 113 /* f2 */ || confirm) {
            this.leaveCell();
            this.cursor.reset();
        }
        else if (isNavigationKey(args.keyCode) && this.cursor.cell.focusGroup.isNavigable()) {
            this.onCursorKeydown(args);
            if (args.defaultPrevented) {
                this.leaveCell();
            }
        }
    };
    NavigationService.prototype.onCellKeydown = function (args) {
        if (this.editService.isEditingCell()) {
            var confirm_1 = args.keyCode === 13 /* enter */;
            var cancel = args.keyCode === 27 /* esc */;
            var navigate = isNavigationKey(args.keyCode);
            if (confirm_1) {
                this.editService.closeCell(args);
            }
            else if (cancel) {
                this.editService.cancelCell();
                this.cd.detectChanges();
            }
            else if (navigate) {
                return false;
            }
        }
        this.cellKeydown.emit(new cell_keydown_event_1.CellKeydownEvent(args));
        return true;
    };
    NavigationService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    NavigationService.ctorParameters = function () { return [
        { type: core_1.NgZone, },
        { type: dom_events_service_1.DomEventsService, },
        { type: pager_context_service_1.PagerContextService, },
        { type: scroll_request_service_1.ScrollRequestService, },
        { type: groups_service_1.GroupsService, },
        { type: focus_root_1.FocusRoot, },
        { type: edit_service_1.EditService, },
        { type: core_1.ChangeDetectorRef, },
        { type: undefined, decorators: [{ type: core_1.Optional }, { type: core_1.Inject, args: [kendo_angular_l10n_1.RTL,] },] },
    ]; };
    return NavigationService;
}());
exports.NavigationService = NavigationService;
