// @progress/kendo-angular-dateinputs/dist/es2015/calendar/services/dom.service.js
import { Injectable } from '@angular/core';
import { CalendarViewEnum } from '../models/view.enum';
import { domContainerFactory as containerFactory, isDocumentAvailable } from '../../util';
const div = containerFactory('div');
const ul = containerFactory('ul');
const li = containerFactory('li');
const td = containerFactory('td');
const th = containerFactory('th');
const tr = containerFactory('tr');
const tbody = containerFactory('tbody');
const thead = containerFactory('thead');
const table = containerFactory('table');
const monthHeader = () => (div(`
            <span class="k-calendar-title">March 2017</span>
            <span class="k-calendar-today">TODAY</span>
        `, 'k-calendar-header'));
const monthWeekHeader = () => (table([
    thead([
        tr([th('MO')])
    ])
], 'k-calendar-weekdays'));
const content = (rows) => (table([
    tbody([
        tr([th('1')]),
        ...(new Array(rows).fill('1').map(v => tr([td(`<span class="k-link">${v}</span>`)])))
    ])
]));
const scrollable = (children) => div(children, 'k-content k-scrollable');
const view = (contentElement, className, renderWeekHeader) => (div([
    monthHeader(),
    renderWeekHeader ? monthWeekHeader() : null,
    scrollable([contentElement, contentElement])
], className, { left: '-10000px', position: 'absolute' }));
const navigationList = (() => {
    let navElement;
    return () => {
        if (!isDocumentAvailable) {
            return null;
        }
        if (!navElement) {
            navElement = div([scrollable([ul([li('<span>FEB</span>')])])], 'k-calendar-navigation', { left: '-10000px', position: 'absolute' });
        }
        return navElement;
    };
})();
const viewFactory = (rows, className, renderWeekHeader) => {
    let viewElement;
    return () => {
        if (!isDocumentAvailable) {
            return null;
        }
        if (!viewElement) {
            viewElement = view(content(rows), className, renderWeekHeader);
        }
        return viewElement;
    };
};
const monthView = viewFactory(6, 'k-calendar-view k-calendar-monthview', true);
const yearView = viewFactory(3, 'k-calendar-view k-calendar-yearview', false);
const decadeView = viewFactory(2, 'k-calendar-view k-calendar-decadeview', false);
const height = (element) => (parseFloat(window.getComputedStyle(element).height) || element.offsetHeight);
const getScrollable = (element) => element.querySelector('.k-scrollable');
const getBody = (element) => element.querySelector('tbody');
/**
 * @hidden
 */
export class DOMService {
    ensureHeights() {
        if (this.calendarHeight !== undefined) {
            return;
        }
        this.calculateHeights();
    }
    calculateHeights(container) {
        if (!isDocumentAvailable()) {
            return;
        }
        this.hostContainer = container;
        this.batch(monthView(), (contentElement) => {
            const viewElement = getBody(contentElement);
            this.calendarHeight = height(contentElement);
            this.monthViewHeight = height(viewElement);
            this.headerHeight = height(viewElement.children[0]);
            this.scrollableContentHeight = height(getScrollable(contentElement));
        });
        this.batch(yearView(), (contentElement) => {
            this.yearViewHeight = height(getBody(contentElement));
            this.scrollableYearContentHeight = height(getScrollable(contentElement));
        });
        this.batch(decadeView(), (contentElement) => {
            this.decadeViewHeight = height(getBody(contentElement));
            this.centuryViewHeight = this.decadeViewHeight;
        });
        this.batch(navigationList(), (contentElement) => {
            this.navigationItemHeight = height(contentElement.querySelector('li'));
        });
    }
    viewHeight(viewType) {
        switch (viewType) {
            case CalendarViewEnum.month:
                return this.monthViewHeight;
            case CalendarViewEnum.year:
                return this.yearViewHeight;
            case CalendarViewEnum.decade:
                return this.decadeViewHeight;
            case CalendarViewEnum.century:
                return this.centuryViewHeight;
            default:
                return 1;
        }
    }
    batch(contentElement, action) {
        const hostContainer = this.hostContainer || document.body;
        const appendedContent = hostContainer.appendChild(contentElement);
        action(appendedContent);
        hostContainer.removeChild(appendedContent);
    }
}
DOMService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
DOMService.ctorParameters = () => [];
