// @progress/kendo-angular-treeview/dist/es/focus-state.service.js
import { Injectable } from "@angular/core";
import { Subject } from "rxjs/Subject";
/**
 * @hidden
 */
var FocusStateService = /** @class */ (function () {
    function FocusStateService() {
        this.changes = new Subject();
        this.activeContent = false;
        this.isActive = false;
    }
    FocusStateService.prototype.hasActiveContent = function () {
        return !this.isActive && this.activeContent;
    };
    FocusStateService.prototype.focus = function (index) {
        this.isActive = true;
        this.changes.next({ focused: true, index: index });
    };
    FocusStateService.prototype.blur = function (index) {
        this.isActive = false;
        this.changes.next({ focused: false, index: index });
    };
    FocusStateService.prototype.focusin = function (_) {
        this.activeContent = true;
    };
    FocusStateService.prototype.focusout = function (_) {
        this.activeContent = false;
    };
    FocusStateService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    FocusStateService.ctorParameters = function () { return []; };
    return FocusStateService;
}());
export { FocusStateService };
